src/workerd/api/streams/writable.c++

```diff
@@ -365,6 +365,154 @@ private:
+
+    return paf.promise.attach(kj::defer([weakRef = weakRef->addRef()]() mutable {
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        auto exception = cancellationException();
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(kj::cp(exception));
+        }
+        KJ_DEFER(obj.writer = kj::none);
+        KJ_IF_SOME(writer, obj.writer) {
+          obj.context.addTask(obj.context.run(
+              [writer=kj::mv(writer), exception=kj::mv(exception)](Worker::Lock& lock) mutable {
+            jsg::Lock& js = lock;
+            auto ex = js.exceptionToJs(kj::mv(exception));
+            return IoContext::current().awaitJs(lock,
+                writer->abort(lock, ex.getHandle(js)));
+          }));
+        }
+      }
+    }));
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    if (size == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run(
+        [ref=weakRef->addRef(), buffer, size](Worker::Lock& lock) mutable
+            -> kj::Promise<void> {
+      jsg::Lock& js = lock;
+      auto& context = IoContext::current();
+      KJ_IF_SOME(obj, ref->tryGet()) {
+        auto source = jsg::BufferSource::wrap(js, const_cast<void*>(buffer), size,
+                                [](void*,size_t,void*) {}, nullptr);
+        v8::Local<v8::Value> value = source.getHandle(js);
+        return context.awaitJs(js, obj.getInner().write(js, value));
+      } else {
+        return cancellationException();
+      }
+    }));
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+
+    auto amount = 0;
+    for (auto& piece : pieces) {
+      amount += piece.size();
+    }
+    if (amount == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run(
+        [ref=weakRef->addRef(), pieces](Worker::Lock& lock) mutable
+            -> kj::Promise<void> {
+      jsg::Lock& js = lock;
+      auto& context = IoContext::current();
+      KJ_IF_SOME(obj, ref->tryGet()) {
+        // This is rather unfortunate because we need to write each piece individually,
+        // which is a separate JS promise :-( ... what we need here is the equivalent of
+        // Promise.all but there is no C++ API provided by v8 to implement that... so we
+        // need to be creative. Still, it's more efficient than individual writes since
+        // we do not need to drop the isolate lock between each piece.
+
+        kj::Maybe<jsg::Promise<void>> lastPromise = kj::none;
+        for (auto& piece : pieces) {
+          if (piece.size() == 0) continue;
+          auto source = jsg::BufferSource::wrap(js, const_cast<byte*>(piece.begin()), piece.size(),
+                                                [](void*,size_t,void*) {}, nullptr);
+          v8::Local<v8::Value> value = source.getHandle(js);
+          auto write = obj.getInner().write(js, value);
+          KJ_IF_SOME(promise, lastPromise) {
```

Comment:
For review: this is the second place in the code that we have code approximating `Promise.all` (here and in socket.c++). We might consider moving it into a jsg utility... or maybe even submitting a patch to v8 to implement a `v8::Promise::All(...)` API.